### ISSUE TITLE
Option for doing all dependency downloads up front before any building happens

### ISSUE THREAD

**Issue by mightybyte** [NixOS/nix#1986]
If you know you are going to be without internet access, it would be useful to have a mechanism for doing all the work that requires an internet connection at the very beginning.  This would allow the user to download all dependencies on short notice and then finish the build process even with no internet access.

**Comment by copumpkin** [comment]
This works today, but doesn't have a dedicated command for it: https://nixos.org/nix-dev/2017-March/023108.html

My main objection to the recommended way is that it involves grepping .drv files for the word `outputHash`, which isn't really a perfect heuristic (since if you just have that string in one of your builds for some reason, it'll also get built). At the same time, I generally prefer small composable operations than big "do what I want" operations with a ton of boolean knobs, so from my perspective I'd rather have a good way to ask Nix if a .drv is FO and then pipe together nix commands rather than have a big precomposed chunk of functionality.

**Comment by 7c6f434c** [comment]
I would also consider `--build-max-jobs 0 --keep-going` or a config file with guaranteed-failing sandbox configuration (again, with `--keep-going` or its abbreviated `-k` form)

**Comment by mightybyte** [comment]
Yeah, I thought about making this ticket be to add a standalone fetch-deps command.  I don't particularly care whether it's a separate command or a command line flag to an existing command.  I just care that it can be done and it's discoverable.

On second thought, there may not need to be any UI at all.  If all the nix commands simply do all the downloading first, then the problem is solved implicitly.

**Comment by lheckemann** [comment]
> If all the nix commands simply do all the downloading first

Then we're also increasing latency on the completion of the build.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
